test/EFCore.PG.FunctionalTests/BulkUpdates/NorthwindBulkUpdatesNpgsqlTest.cs

```diff
@@ -659,20 +990,248 @@ public override async Task Update_with_invalid_lambda_in_set_property_throws(boo
+    FROM ""Customers"" AS c1
+    WHERE c1.""CustomerID"" LIKE 'A%'
+) AS t
+WHERE c.""CustomerID"" = t.""CustomerID""");
+    }
+
+    public override async Task Update_Intersect_set_constant(bool async)
+    {
+        await base.Update_Intersect_set_constant(async);
+
+        AssertExecuteUpdateSql(
+            @"UPDATE ""Customers"" AS c
+    SET ""ContactName"" = 'Updated'
+FROM (
+    SELECT c0.""CustomerID"", c0.""Address"", c0.""City"", c0.""CompanyName"", c0.""ContactName"", c0.""ContactTitle"", c0.""Country"", c0.""Fax"", c0.""Phone"", c0.""PostalCode"", c0.""Region""
+    FROM ""Customers"" AS c0
+    WHERE c0.""CustomerID"" LIKE 'F%'
+    INTERSECT
+    SELECT c1.""CustomerID"", c1.""Address"", c1.""City"", c1.""CompanyName"", c1.""ContactName"", c1.""ContactTitle"", c1.""Country"", c1.""Fax"", c1.""Phone"", c1.""PostalCode"", c1.""Region""
+    FROM ""Customers"" AS c1
+    WHERE c1.""CustomerID"" LIKE 'A%'
+) AS t
+WHERE c.""CustomerID"" = t.""CustomerID""");
+    }
+
+    public override async Task Update_with_join_set_constant(bool async)
+    {
+        await base.Update_with_join_set_constant(async);
+
+        AssertExecuteUpdateSql(
+            @"UPDATE ""Customers"" AS c
+    SET ""ContactName"" = 'Updated'
+FROM (
+    SELECT o.""OrderID"", o.""CustomerID"", o.""EmployeeID"", o.""OrderDate""
+    FROM ""Orders"" AS o
+    WHERE o.""OrderID"" < 10300
+) AS t
+WHERE c.""CustomerID"" = t.""CustomerID"" AND (c.""CustomerID"" LIKE 'F%')");
+    }
+
+    public override async Task Update_with_left_join_set_constant(bool async)
+    {
+        await base.Update_with_left_join_set_constant(async);
+
+        AssertExecuteUpdateSql(
+            @"UPDATE ""Customers"" AS c
+    SET ""ContactName"" = 'Updated'
+FROM (
+    SELECT c0.""CustomerID"", c0.""Address"", c0.""City"", c0.""CompanyName"", c0.""ContactName"", c0.""ContactTitle"", c0.""Country"", c0.""Fax"", c0.""Phone"", c0.""PostalCode"", c0.""Region"", t.""OrderID"", t.""CustomerID"" AS ""CustomerID0"", t.""EmployeeID"", t.""OrderDate""
+    FROM ""Customers"" AS c0
+    LEFT JOIN (
+        SELECT o.""OrderID"", o.""CustomerID"", o.""EmployeeID"", o.""OrderDate""
+        FROM ""Orders"" AS o
+        WHERE o.""OrderID"" < 10300
+    ) AS t ON c0.""CustomerID"" = t.""CustomerID""
+    WHERE c0.""CustomerID"" LIKE 'F%'
+) AS t0
+WHERE c.""CustomerID"" = t0.""CustomerID""");
+    }
+
+    public override async Task Update_with_cross_join_set_constant(bool async)
+    {
+        await base.Update_with_cross_join_set_constant(async);
+
+        AssertExecuteUpdateSql(
+            @"UPDATE ""Customers"" AS c
+    SET ""ContactName"" = 'Updated'
+FROM (
+    SELECT o.""OrderID"", o.""CustomerID"", o.""EmployeeID"", o.""OrderDate""
+    FROM ""Orders"" AS o
+    WHERE o.""OrderID"" < 10300
+) AS t
+WHERE c.""CustomerID"" LIKE 'F%'");
+    }
+
+    [ConditionalTheory(Skip = "invalid reference to FROM-clause entry for table c")]
```

Comment:
/cc @smitpatel

Reply:
What is invalid reference here? Sqlite doesn't have APPLY support so this test is skipped. May be we can tweak relational a bit here.

Reply:
Update_with_cross_apply_set_constant is producing the following SQL:

```sql
UPDATE "Customers" AS c
    SET "ContactName" = 'Updated'
FROM (
    SELECT o."OrderID", o."CustomerID", o."EmployeeID", o."OrderDate"
    FROM "Orders" AS o
    WHERE o."OrderID" < 10300 AND date_part('year', o."OrderDate")::int < length(c."ContactName")::int
) AS t
WHERE c."CustomerID" LIKE 'F%'
```

The error is:

```
ERROR:  invalid reference to FROM-clause entry for table "c" at character 240
HINT:  There is an entry for table "c", but it cannot be referenced from this part of the query.
```

So this is like the problem we discussed offline with LEFT JOIN - not possible to reference c here in the subquery, I think maybe you need to push it down into a nested subquery like for the LEFT JOIN case?

Reply:
That is odd to not allow reference. I will update relational code to cause inner join.

Reply:
@smitpatel just remembered this, do you want me to file an issue to track it or are we like this?

Reply:
yes, please file an issue. Thinking about this more, I think it is legit that it doesn't work. I have more ideas around heuristics on this. Will add those notes on the issue you file so we have it for reference in future.

Reply:
#28823